engineio/transport/polling/decoder.go

```diff
@@ -0,0 +1,128 @@
+package polling
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"github.com/googollee/go-socket.io/engineio/frame"
+)
+
+// decoder decodes frames from the reader.
+type decoder struct {
+	reader    *bufReader
+	lastFrame *frameReader
+}
+
+func newDecoder(buf []byte, r io.Reader) *decoder {
+	return &decoder{
+		reader: newBufReader(buf, r),
+	}
+}
+
+// NextFrame returns a reader to read the next frame.
+func (d *decoder) NextFrame() (frame.Type, io.Reader, error) {
+	if d.lastFrame != nil {
+		if err := d.lastFrame.Discard(); err != nil {
+			return 0, nil, err
+		}
+	}
+
+	if err := d.reader.Fill(); err != nil {
+		return 0, nil, err
+	}
+
+	next, err := d.reader.ReadByte()
+	if err != nil && err != io.EOF {
+		return 0, nil, err
+	}
+
+	if d.lastFrame != nil {
+		// Start from the 2nd frame, need to ignore the separator.
+		next, err = d.reader.ReadByte()
+		if err != nil && err != io.EOF {
+			return 0, nil, err
+		}
+	}
+
+	d.lastFrame = &frameReader{
+		reader:   d.reader,
+		finished: false,
+	}
+
+	if err == nil && next == binaryPrefix {
+		return frame.Binary, base64.NewDecoder(base64.StdEncoding, d.lastFrame), nil
+	}
+
+	// The next byte is not binary prefix byte. Push it back to the reader
+	if err == nil {
+		if err := d.lastFrame.reader.PushBack(1); err != nil {
+			return 0, nil, err
+		}
+	}
+
+	return frame.Text, d.lastFrame, nil
+}
+
+// frameReader is a reader to read one frame.
+type frameReader struct {
```

Comment:
i suggest, we need separate `frameReader` and encoder by two different file in one package

Reply:
I did before, but as `decoder` and `frameReader` are so closed to each other, I need to switch between two files often. So I put them together.